itest/rust/src/engine_tests/native_structures_test.rs

```diff
@@ -70,6 +75,419 @@ impl ITextServerExtension for TestTextServer {
+        _fonts: Array<Rid>,
+        _size: i64,
+        _opentype_features: Dictionary,
+        _language: GString,
+        _meta: Variant,
+    ) -> bool {
+        unreachable!()
+    }
+
+    fn shaped_text_add_object(
+        &mut self,
+        _shaped: Rid,
+        _key: Variant,
+        _size: Vector2,
+        _inline_align: InlineAlignment,
+        _length: i64,
+        _baseline: f64,
+    ) -> bool {
+        unreachable!()
+    }
+
+    fn shaped_text_resize_object(
+        &mut self,
+        _shaped: Rid,
+        _key: Variant,
+        _size: Vector2,
+        _inline_align: InlineAlignment,
+        _baseline: f64,
+    ) -> bool {
+        unreachable!()
+    }
+
+    fn shaped_get_span_count(&self, _shaped: Rid) -> i64 {
+        unreachable!()
+    }
+
+    fn shaped_get_span_meta(&self, _shaped: Rid, _index: i64) -> Variant {
+        unreachable!()
+    }
+
+    fn shaped_set_span_update_font(
+        &mut self,
+        _shaped: Rid,
+        _index: i64,
+        _fonts: Array<Rid>,
+        _size: i64,
+        _opentype_features: Dictionary,
+    ) {
+        unreachable!()
+    }
+
+    fn shaped_text_substr(&self, _shaped: Rid, _start: i64, _lengthh: i64) -> Rid {
+        unreachable!()
+    }
+
+    fn shaped_text_get_parent(&self, _shaped: Rid) -> Rid {
+        unreachable!()
+    }
+
+    fn shaped_text_shape(&mut self, _shaped: Rid) -> bool {
+        unreachable!()
+    }
+
+    fn shaped_text_is_ready(&self, _shaped: Rid) -> bool {
+        unreachable!()
+    }
+
+    unsafe fn shaped_text_sort_logical(&mut self, _shaped: Rid) -> *const Glyph {
+        unreachable!()
+    }
+
+    fn shaped_text_get_range(&self, _shaped: Rid) -> Vector2i {
+        unreachable!()
+    }
+
+    fn shaped_text_get_trim_pos(&self, _shaped: Rid) -> i64 {
+        unreachable!()
+    }
+
+    fn shaped_text_get_ellipsis_pos(&self, _shaped: Rid) -> i64 {
+        unreachable!()
+    }
+
+    fn shaped_text_get_ellipsis_glyph_count(&self, _shaped: Rid) -> i64 {
+        unreachable!()
+    }
+
+    unsafe fn shaped_text_get_ellipsis_glyphs(&self, _shaped: Rid) -> *const Glyph {
+        unreachable!()
+    }
+
+    fn shaped_text_get_objects(&self, _shaped: Rid) -> VariantArray {
+        unreachable!()
+    }
+
+    fn shaped_text_get_object_rect(&self, _shaped: Rid, _key: Variant) -> Rect2 {
+        unreachable!()
+    }
+
+    fn shaped_text_get_size(&self, _shaped: Rid) -> Vector2 {
+        unreachable!()
+    }
+
+    fn shaped_text_get_ascent(&self, _shaped: Rid) -> f64 {
+        unreachable!()
+    }
+
+    fn shaped_text_get_descent(&self, _shaped: Rid) -> f64 {
+        unreachable!()
+    }
+
+    fn shaped_text_get_width(&self, _shaped: Rid) -> f64 {
+        unreachable!()
+    }
+
+    fn shaped_text_get_underline_position(&self, _shaped: Rid) -> f64 {
+        unreachable!()
+    }
+
+    fn shaped_text_get_underline_thickness(&self, _shaped: Rid) -> f64 {
+        unreachable!()
+    }
```

Comment:
This adds 400 lines of useless code, which has nothing to do with the native structures. The surrounding tests were working before this addition, so objectively they're not _always_ required.

This makes me question if we should really enforce overriding all virtual methods. Maybe there could be an opt-out or so...
```rs
#[godot_api(no_required_virtual)]
```

But please let's discuss this a bit more before jumping to implementation :slightly_smiling_face:

Reply:
> One thing we don't know is whether "required" always implies "is indeed called". Because in all the cases where a required virtual method is not called, we now impose useless code on the user.

To address this, from what I understand, required does not imply "is indeed called" but "must be implemented when it is called". This means that Godot will not guarantee to use some default value if the method is not implemented. Occasionally, it's ok to ignore required methods, but often it can break the editor / engine or crash it entirely.

So if we want to provide defaults for these methods, we have to get the `godot_api` macro to generate all the missing methods. We can't do this just with the trait implementation and need either the `ExtensionApi` (means reloading the `extension_api.json` for every macro invocation... sounds unreasonable to me), generate the macro code in codegen, that can then generate the default implementation if needed (sounds complex to follow and reason about) or hand implement all the defaults inside the macro (probably quite tedious).

We can also leave the default implementation inside the generated trait definition, but we still need a call inside `__virtual_call` to all the required methods and the macro doesn't know which methods are required.

If you have other ideas, let me know.

Reply:
> generate the macro code in codegen, that can then generate the default implementation if needed (sounds complex to follow and reason about)

We do this for the `unsafe_inherits_transitive_*` macros. It's definitely possible, but it causes codegen to emit quite a bit of extra code that may or may not be useful for the average user.

---

> We can also leave the default implementation inside the generated trait definition, but we still need a call inside `__virtual_call` to all the required methods and the macro doesn't know which methods are required.

Can you elaborate this?

---



> Yes, they are technically not always required, but they are required to get a working `TextServer` implementation. Using a `TextServer` that does not implement some of these methods _might_ break the engine (at least that was my experience with 'ScriptLanguageExtension`). It's unpredictable.

I would also err on the conservative side, however there are likely methods that are required but not necessarily called at runtime (also shown by the test, even if that's not typical gamedev code). If we force users to have dozens of function bodies containing `todo!()`, we don't achieve very much.

Reply:
The problem that tests are now bloated with 400 useless LoC (and 55 in another instance) remains though. This already caused a lot of problems:
- It requires extra dependencies (`Material` in this case), slowing down CI. 
- It increases maintenance effort, if:
   - Godot adds new required methods (technically breaking, but possible).
   - Godot changes signatures of something marked experimental (not sure if applicable).
   - We change the way how the code is generated, e.g. [#649](https://github.com/godot-rust/gdext/issues/649).
- Last but not least, it's 400 LoC that have no meaning for the test logic but are still there.

If we don't want to go the whole route of "allow default implementations failing at runtime", we could see if we can use another class than `TextServerExtension` here, to check native structure interaction. Or to approach this differently somehow.

Last resort would be to generate different code just for tests, but I'd rather not go this route unless absolutely necessary.

Reply:
Okay, that makes the priorities more clear to me. 

I looked into the problem a bit and unfortunately, there is not really a better option for native structure testing. There are not many native structures, and they either relate to the Physics Servers, Text server, script language or audio processing.  They all contain a lot of required methods or don't have a public interface that would allow testing them. Switching the class does not seem feasible so far. 

I haven't had a good idea yet how we could solve this excessive boilerplate in the tests in a convenient way.